src/tool_operate.c

```diff
@@ -144,6 +155,29 @@ static bool is_pkcs11_uri(const char *string)
   }
 }
 
+static int sockopt_callback(void *clientp, curl_socket_t curlfd,
+                            curlsocktype purpose)
+{
+  struct sockopt_data *val = (struct sockopt_data *)clientp;
+  if(purpose != CURLSOCKTYPE_IPCXN)
+    return CURL_SOCKOPT_OK;
+  (void)val;
+  (void)curlfd;
+#ifdef IP_TOS
+  if(val->ip_tos > 0 && val->ip_tos <= 0xff) {
+    int tos = (int)val->ip_tos;
+    if(setsockopt(curlfd, SOL_IP, IP_TOS,
+      (const char *)&tos, sizeof(tos)) < 0) {
+      int error = errno;
+      fprintf(tool_stderr, "IP_TOS %d failed with errno %d: %s;\n",
+              tos, error, strerror(error));
```

Comment:
Can you suggest a way to check the protocol type here?

Reply:
Can `getsockname` help?

Reply:
Actually I found `getsockopt(sockfd, SOL_SOCKET, SO_DOMAIN, &family, &len)`.

But before I change that, isn't it better to show an error on such failure rather than silently ignoring the option?

Reply:
This goes back to what I asked you from the start: users typically won't know if IPv4 or IPv6 gets used so they can't tell when this option takes effect. Do you think a user wants an error message when it doesn't? I doubt that.

Since v4/v6 may even vary between separate invokes to the same host name it could even result in an error message intermittently every nth invoke which probably will be highly confusing and hardly productive for users.

Reply:
Done